### PR TITLE
GH-192 Python release integration test

### PR DESCRIPTION
**Pull request by velzevur**
All builds with a tag set to anything will trigger a Tavis JOB (one per OS - linux and osx) to validate the release is correct. The script does the following

- generate a release like tarball (only change being fast mining is enabled)
- extract the tarball
- copy the extracted dir to 2 more directories (so we have node1, node2, node3)
- setup the nodes
- start the nodes
- get each node's top every second untill there are 50 blocks mined
- stop the nodes and cleanup

If a node crashes - all node logs are tailed so we can easily debug it.

It is a huge commit and it is mostly due to that I had to add some Swagger generated python code (so we don't have to install swager-codegen in the JOB). The generated code is located in ` py/tests/swagger_client/ `
You can focus on changes in `.travis.yml`, `Makefile`, `ci`, `py/tests/release.py` and `rebar.config`

### CODE REVIEW

**Comment by lucafavatella** [comment]
> TODO: check what's the real value of TRAVIS_TAG when building a release

@velzevur Wrt checking this for the case the build is a tag: I recommend you try this by forking this repo in your github account, enabling Travis CI there, pushing this branch there too, taking there (in your fork) a tag, observing result.

File: py/tests/release.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+from swagger_client.api_client import ApiClient
+
+def build_dir():
+    return "build_dir"
+
+def executable():
+    return os.path.join(build_dir(), "bin", "epoch")
+
+def extract_tarball(tarball_name):
+    extract_dir = build_dir()
```
Comment (lucafavatella) [inline py/tests/release.py]: Please do not change the build dir. Travis uploads a file from the build dir (see `file_glob: true` and `file: _build/prod/rel/epoch/*-epoch-0.1.0.tar.gz` in .travis.yml) and extracting stuff in the build dir may confuse the upload by Travis. Please use a temporary folder for extracting tarball.

Comment (lucafavatella) [reply py/tests/release.py]: BTW that file path in travis.yml shall be made more dynamic e.g. `_build/prod/rel/epoch/*-epoch-*.tar.gz` (untested).

Comment (velzevur) [reply py/tests/release.py]: Yes, it should be

File: ci [hunk 2]
```diff
@@ -58,7 +58,17 @@ case "${1:?}"-"${2:?}" in
         BuildDir="${3:?}"
         ( cd "${BuildDir:?}" && ./rebar3 as prod tar; )
         Prefix=$(package_prefix "${TRAVIS_OS_NAME:?}")
-        cp -p "${BuildDir:?}"/_build/prod/rel/epoch/epoch-0.2.0.tar.gz "${BuildDir:?}"/_build/prod/rel/epoch/"${Prefix:?}"-epoch-0.2.0.tar.gz
+        cp -p "${BuildDir:?}"/_build/prod/rel/epoch/epoch-0.1.0.tar.gz "${BuildDir:?}"/_build/prod/rel/epoch/"${Prefix:?}"-epoch-0.1.0.tar.gz
```
Comment (lucafavatella) [inline ci]: 0.1.0 or 0.2.0?

File: rebar.config [hunk 3]
```diff
@@ -10,7 +10,7 @@
         {jobs, "0.7.0"}
        ]}.
 
-{relx, [{release, { epoch, "0.2.0" },
+{relx, [{release, { epoch, "0.1.0" },
```
Comment (lucafavatella) [inline rebar.config]: 0.1.0 or 0.2.0?

File: ci [hunk 4]
```diff
@@ -58,7 +58,17 @@ case "${1:?}"-"${2:?}" in
...
+        ;;
+    script-acceptance-test)
```
Comment (lucafavatella) [inline ci]: I see the block of code below is not hooked to the CI. Is this meant to be a new job (see .travis.yml) or is this meant to be executed in the `package` job (see .travis.yml)?

File: ci (hunk 2 above)
Comment (velzevur) [reply ci]: ah! Nice catch, thanks

File: ci [hunk 5]
```diff
@@ -60,6 +60,26 @@ case "${1:?}"-"${2:?}" in
...
+        if [ "$TRAVIS_TAG" != "" ]; then
+          BuildDir="${3:?}"
+          ( cd "${BuildDir:?}" && ./rebar3 as fastmining tar; )
+          make nose-env
+          make release-integration-test TARBALL=_build/fastmining/rel/epoch/epoch-0.2.0.tar.gz
+        else
+          true
+        fi
+        ;;
```
Comment (lucafavatella) [inline ci]: This section looks repeated twice.

File: rebar.config [hunk 6]
```diff
@@ -38,6 +38,11 @@
                              {vm_args, "./config/dev1/vm.args"}]}
                     ]
             },
+            {fastmining, [{relx, [{dev_mode, false},
```
Comment (lucafavatella) [inline rebar.config]: I understand this profile will be deleted once the production release has mining speed configurable, right?
(no need to amend PR for this - just checking we have same understanding)

Comment (velzevur) [reply rebar.config]: Yes, exactly

File: ci [hunk 7]
```diff
@@ -60,6 +60,12 @@ case "${1:?}"-"${2:?}" in
         Prefix=$(package_prefix "${TRAVIS_OS_NAME:?}")
         cp -p "${BuildDir:?}"/_build/prod/rel/epoch/epoch-0.2.0.tar.gz "${BuildDir:?}"/_build/prod/rel/epoch/"${Prefix:?}"-epoch-0.2.0.tar.gz
         ;;
+    script-acceptance-test)
+        BuildDir="${3:?}"
+        ( cd "${BuildDir:?}" && ./rebar3 as fastmining tar; )
+        make nose-env
+        make release-integration-test TARBALL=_build/fastmining/rel/epoch/epoch-0.2.0.tar.gz
```
Comment (lucafavatella) [inline ci]: I see master may have moved on from the hardcoded 0.2.0.

File: py/tests/release.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+import swagger_client
+from swagger_client.rest import ApiException
+from swagger_client.apis.external_api import ExternalApi
+from swagger_client.api_client import ApiClient
+
+SETUP = {
+        "node1": {
+            "host": "localhost:9900/v1",
+            "commands": [
+                "sed -ibkp 's/-sname epoch/-sname epoch1/g' ./releases/0.2.0/vm.args"
```
Comment (lucafavatella) [inline py/tests/release.py]: Again - master may have moved on from hardcoded 0.2.0

File: py/tests/release.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+ [
+  {swagger_port_external, 9900}
+ ]}
+].
+'''
+                },
+        "node2": {
+            "host": "localhost:9902/v1",
+            "commands": [
+                "sed -ibkp 's/-sname epoch/-sname epoch2/g' ./releases/0.2.0/vm.args"
```
Comment (lucafavatella) [inline py/tests/release.py]: same

File: py/tests/release.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+ [
+  {swagger_port_external, 9902}
+ ]}
+].
+'''
+                },
+        "node3": {
+            "host": "localhost:9903/v1",
+            "commands": [
+                "sed -ibkp 's/-sname epoch/-sname epoch3/g' ./releases/0.2.0/vm.args"
```
Comment (lucafavatella) [inline py/tests/release.py]: same

**Pull request closed without merging**